pkg/machine/machineset.go

```diff
@@ -0,0 +1,538 @@
+		glog.V(100).Infof("The replicas of the MachineSet is zero")
+
+		builder.errorMsg = "MachineSet 'replicas' cannot be zero"
+	}
+
+	if builder.Definition == nil {
+		glog.V(100).Infof("The MachineSet object definition is nil")
+
+		builder.errorMsg = "MachineSet 'Object.Definition' is nil"
+	}
+
+	return &builder
+}
+
+// getPublicCloudKind determines the public cloud kind and stores it in the builder struct.
+func (builder *MachineSetBuilder) getPublicCloudKind() error {
+	if valid, err := builder.validate(); !valid {
+		return err
+	}
+
+	glog.V(100).Infof("Determining the public cloud kind")
+
+	providerSpecMap := make(map[string]interface{})
+
+	// Value field is of type *runtime.RawExtension
+	byteArray, err := json.Marshal(builder.Definition.Spec.Template.Spec.ProviderSpec.Value)
+
+	if err != nil {
+		builder.errorMsg = fmt.Sprintf("error determining public cloud kind: %v", err)
+
+		return fmt.Errorf("error marshalling the providerSpec Value element into a byte array")
+	}
+
+	err = json.Unmarshal(byteArray, &providerSpecMap)
+
+	if err != nil {
+		builder.errorMsg = fmt.Sprintf("error determining public cloud kind: %v", err)
+
+		return fmt.Errorf("error unmarshalling the byte array into a providerSpec map")
+	}
+
+	// publicCloudKind is of type: map[string]interface{}
+	publicCloudKind := providerSpecMap["kind"]
+
+	publicCloud, ok := publicCloudKind.(string)
+
+	if !ok {
+		return fmt.Errorf("failed to detect public cloud kind")
+	}
+
+	glog.V(100).Infof("ProviderSpec kind param is '%s'", publicCloud)
+
+	switch publicCloud {
+	case "AWSMachineProviderConfig":
+		builder.publicCloud = "aws"
+	case "GCPMachineProviderSpec":
+		builder.publicCloud = "gcp"
+	case "AzureMachineProviderSpec":
+		builder.publicCloud = "azure"
+	default:
+		builder.errorMsg = "unsupported cloud platform. Supported public cloud are AWS, GCP, and Azure"
+
+		return fmt.Errorf("unsupported cloud platform. Supported public cloud are AWS, GCP, and Azure")
+	}
+
+	return nil
+}
+
+// PullMachineSet loads an existing MachineSet into Builder struct.
+func PullMachineSet(apiClient *clients.Settings, name, namespace string) (*MachineSetBuilder, error) {
+	glog.V(100).Infof("Pulling existing machineSet name %s in namespace %s", name, namespace)
+
+	builder := MachineSetBuilder{
+		apiClient: apiClient,
+		Definition: &machinev1beta1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		},
+	}
+
+	if name == "" {
+		builder.errorMsg = "MachineSet 'name' cannot be empty"
+	}
+
+	if namespace == "" {
+		builder.errorMsg = "MachineSet 'namespace' cannot be empty"
+	}
+
+	if !builder.Exists() {
+		return nil, fmt.Errorf("machineSet object %s doesn't exist in namespace %s", name, namespace)
+	}
+
+	builder.Definition = builder.Object
+
+	return &builder, nil
+}
+
+// createNewWorkerMachineSetFromCopy returns a MachineSetBuilder object which was copied from an exiting one on cluster.
+func createNewWorkerMachineSetFromCopy(
```

Comment:
Please move it down to file. Because it private function.

Reply:
Done